Source/missiles.cpp

```diff
@@ -261,113 +261,130 @@ int GetSpellLevel(int id, int sn)
 	return result;
 }
 
+/**
+ * @brief Returns the direction a vector from p1(x1, y1) to p2(x2, y2) is pointing to.
+ *
+ *      W    SW     S
+ *            ^
+ *            |	
+ *     NW ----+---> SE
+ *            |
+ *            |
+ *      N    NE     E
+ *
+ * @param x1 the x coordinate of p1
+ * @param y1 the y coordinate of p1
+ * @param x2 the x coordinate of p2
+ * @param y2 the y coordinate of p2
+ * @return the direction of the p1->p2 vector
+*/
 int GetDirection8(int x1, int y1, int x2, int y2)
 {
-	BYTE Dirs[16][16] = {
-		{ 99, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
-		{ 2, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
-		{ 2, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
-		{ 2, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0 },
-		{ 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0 },
-		{ 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0 },
-		{ 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 },
-		{ 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 },
-		{ 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 },
-		{ 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 },
-		{ 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 },
-		{ 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 },
-		{ 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 },
-		{ 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 },
-		{ 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 },
-		{ 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 }
-	};
 	int mx, my, md;
-	ALIGN_BY_1 BYTE urtoll[] = { 3, 4, 5 },
-	                ultolr[] = { 3, 2, 1 },
-	                lrtoul[] = { 7, 6, 5 },
-	                lltour[] = { 7, 0, 1 };
-
-	mx = abs(x2 - x1);
-	if (mx > 15)
-		mx = 15;
-	my = abs(y2 - y1);
-	if (my > 15)
-		my = 15;
-	md = Dirs[my][mx];
-	if (x1 > x2) {
-		if (y1 > y2)
-			md = urtoll[md];
-		else
-			md = ultolr[md];
-	} else if (y1 > y2)
-		md = lrtoul[md];
-	else
-		md = lltour[md];
+
+	mx = x2 - x1;
+	my = y2 - y1;
+	if (mx >= 0) {
+		if (my >= 0) {
+			if (5 * mx <= (my << 1)) // mx/my <= 0.4, approximation of tan(22.5)
+				return 1; // DIR_SW
+			md = 0; // DIR_S
+		} else {
+			my = -my;
+			if (5 * mx <= (my << 1))
+				return 5; // DIR_NE
+			md = 6; // DIR_E
+		}
+		if (5 * my <= (mx << 1))         // my/mx <= 0.4
+			md = 7; // DIR_SE
+	} else {
+		mx = -mx;
+		if (my >= 0) {
+			if (5 * mx <= (my << 1))
+				return 1; // DIR_SW
+			md = 2; // DIR_W
+		} else {
+			my = -my;
+			if (5 * mx <= (my << 1))
+				return 5; // DIR_NE
+			md = 4; // DIR_N
+		}
+		if (5 * my <= (mx << 1))
+			md = 3; // DIR_NW
+	}

```

Comment:
Seems to me like the `if` block and the `else` block are pretty much identical. Is there no way to share the logic here?

I see the original code was using `abs`. Can't we do something similar to avoid having to duplicate all the checks?

Reply:
@pionere did clean it up, but never updated the PR so some one will  have to go dig it out:
https://github.com/pionere/devilutionX/blob/master/Source/missiles.cpp#L285

Reply:
@pionere has made a more minimalistic version, but some one would have to go dig it out:
https://github.com/pionere/devilutionX/blob/master/Source/missiles.cpp#L285